File: spark/wallet/deposit.go
package wallet

import (
	"bytes"
	"context"
	"fmt"

	"github.com/btcsuite/btcd/btcutil"
	"github.com/btcsuite/btcd/txscript"
	"github.com/btcsuite/btcd/wire"
	"github.com/decred/dcrd/dcrec/secp256k1"
	"github.com/google/uuid"
	"github.com/lightsparkdev/spark-go/common"
	pbcommon "github.com/lightsparkdev/spark-go/proto/common"
	pbfrost "github.com/lightsparkdev/spark-go/proto/frost"
	pb "github.com/lightsparkdev/spark-go/proto/spark"
)

// GenerateDepositAddress generates a deposit address for a given identity and signing public key.
func GenerateDepositAddress(
	ctx context.Context,
	config *Config,
	identityPubkey, signingPubkey []byte,
) (*pb.GenerateDepositAddressResponse, error) {
	sparkConn, err := common.NewGRPCConnection(config.SparkServiceAddress)
	if err != nil {
		return nil, err
	}
	defer sparkConn.Close()
	sparkClient := pb.NewSparkServiceClient(sparkConn)
	depositResp, err := sparkClient.GenerateDepositAddress(ctx, &pb.GenerateDepositAddressRequest{
		SigningPublicKey:  signingPubkey,
		IdentityPublicKey: identityPubkey,
	})
	if err != nil {
		return nil, err
	}
	return depositResp, nil
}

// CreateTree creates a tree for a given deposit transaction.
func CreateTree(
	ctx context.Context,
	config *Config,
	identityPubkey,
	signingPrivKey,
	verifyingKey []byte,
	depositTx *wire.MsgTx,
	vout int,
) (*pb.FinalizeNodeSignaturesResponse, error) {
	_, signingPubkey := secp256k1.PrivKeyFromBytes(signingPrivKey)
	signingPubkeyBytes := signingPubkey.SerializeCompressed()
	// Creat root tx
	rootTx := wire.NewMsgTx(2)
	rootTx.AddTxIn(wire.NewTxIn(
		&wire.OutPoint{Hash: depositTx.TxHash(), Index: uint32(vout)},
		depositTx.TxOut[0].PkScript,
		nil, // witness
	))
	rootTx.AddTxOut(wire.NewTxOut(100_000, depositTx.TxOut[0].PkScript))
	var rootBuf bytes.Buffer
	rootTx.Serialize(&rootBuf)
	rootNonceHidingPriv, _ := secp256k1.GeneratePrivateKey()
	rootNonceBidingPriv, _ := secp256k1.GeneratePrivateKey()
	rootNonceCommitment := pbcommon.SigningCommitment{
		Hiding:  rootNonceHidingPriv.PubKey().SerializeCompressed(),
		Binding: rootNonceBidingPriv.PubKey().SerializeCompressed(),
	}
	rootTxSighash, err := common.SigHashFromTx(rootTx, 0, depositTx.TxOut[0])
	if err != nil {
		return nil, err
	}

	// Creat refund tx
	refundTx := wire.NewMsgTx(2)
	refundTx.AddTxIn(wire.NewTxIn(
		&wire.OutPoint{Hash: rootTx.TxHash(), Index: 0},
		rootTx.TxOut[0].PkScript,
		nil, // witness
	))
	refundP2trAddress, _ := common.P2TRAddressFromPublicKey(signingPubkeyBytes, config.Network)
	refundAddress, _ := btcutil.DecodeAddress(*refundP2trAddress, common.NetworkParams(config.Network))
	refundPkScript, _ := txscript.PayToAddrScript(refundAddress)
	refundTx.AddTxOut(wire.NewTxOut(100_000, refundPkScript))
	refundTx.LockTime = 60000
	var refundBuf bytes.Buffer
	refundTx.Serialize(&refundBuf)
	refundNonceHidingPriv, _ := secp256k1.GeneratePrivateKey()
	refundNonceBidingPriv, _ := secp256k1.GeneratePrivateKey()
	refundNonceCommitment := pbcommon.SigningCommitment{
		Hiding:  refundNonceHidingPriv.PubKey().SerializeCompressed(),
		Binding: refundNonceBidingPriv.PubKey().SerializeCompressed(),
	}
	refundTxSighash, err := common.SigHashFromTx(refundTx, 0, rootTx.TxOut[0])
	if err != nil {
		return nil, err
	}

	sparkConn, err := common.NewGRPCConnection(config.SparkServiceAddress)
	if err != nil {
		return nil, err
	}
	defer sparkConn.Close()
	sparkClient := pb.NewSparkServiceClient(sparkConn)

	treeResponse, err := sparkClient.StartTreeCreation(ctx, &pb.StartTreeCreationRequest{
		IdentityPublicKey: identityPubkey,
		OnChainUtxo: &pb.UTXO{
			Txid: depositTx.TxID(),
			Vout: uint32(vout),
		},
		RootTxSigningJob: &pb.SigningJob{
			RawTx:                  rootBuf.Bytes(),
			SigningPublicKey:       signingPubkeyBytes,
			SigningNonceCommitment: &rootNonceCommitment,
		},
		RefundTxSigningJob: &pb.SigningJob{
			RawTx:                  refundBuf.Bytes(),
			SigningPublicKey:       signingPubkeyBytes,
			SigningNonceCommitment: &refundNonceCommitment,
		},
	})
	if err != nil {
		return nil, err
	}

	if !bytes.Equal(treeResponse.RootNodeSignatureShares.VerifyingKey, verifyingKey) {
		return nil, fmt.Errorf("verifying key does not match")
	}

	userIdentifier := "0000000000000000000000000000000000000000000000000000000000000063"
	userKeyPackage := pbfrost.KeyPackage{
		Identifier:  userIdentifier,
		SecretShare: signingPrivKey,
		PublicShares: map[string][]byte{
			userIdentifier: signingPubkeyBytes,
		},
		PublicKey:  treeResponse.RootNodeSignatureShares.VerifyingKey,
		MinSigners: 1,
	}

	userSigningJobs := make([]*pbfrost.FrostSigningJob, 0)
	nodeJobID := uuid.NewString()
	refundJobID := uuid.NewString()
	userSigningJobs = append(userSigningJobs, &pbfrost.FrostSigningJob{
		JobId:        nodeJobID,
		Message:      rootTxSighash,
		KeyPackage:   &userKeyPackage,
		VerifyingKey: verifyingKey,
		Nonce: &pbfrost.SigningNonce{
			Hiding:  rootNonceHidingPriv.Serialize(),
			Binding: rootNonceBidingPriv.Serialize(),
		},
		Commitments:     treeResponse.RootNodeSignatureShares.NodeTxSigningResult.SigningNonceCommitments,
		UserCommitments: &rootNonceCommitment,
	})
	userSigningJobs = append(userSigningJobs, &pbfrost.FrostSigningJob{
		JobId:        refundJobID,
		Message:      refundTxSighash,
		KeyPackage:   &userKeyPackage,
		VerifyingKey: treeResponse.RootNodeSignatureShares.VerifyingKey,
		Nonce: &pbfrost.SigningNonce{
			Hiding:  refundNonceHidingPriv.Serialize(),
			Binding: refundNonceBidingPriv.Serialize(),
		},
		Commitments:     treeResponse.RootNodeSignatureShares.RefundTxSigningResult.SigningNonceCommitments,
		UserCommitments: &refundNonceCommitment,
	})

	frostConn, err := common.NewGRPCConnection(config.FrostSignerAddress)
	if err != nil {
		return nil, err
	}
	defer frostConn.Close()

	frostClient := pbfrost.NewFrostServiceClient(frostConn)

	userSignatures, err := frostClient.SignFrost(context.Background(), &pbfrost.SignFrostRequest{
		SigningJobs: userSigningJobs,
		Role:        pbfrost.SigningRole_USER,
	})
	if err != nil {
		return nil, err
	}

	rootSignature, err := frostClient.AggregateFrost(context.Background(), &pbfrost.AggregateFrostRequest{
		Message:            rootTxSighash,
		SignatureShares:    treeResponse.RootNodeSignatureShares.NodeTxSigningResult.SignatureShares,
		PublicShares:       treeResponse.RootNodeSignatureShares.NodeTxSigningResult.PublicKeys,
		VerifyingKey:       verifyingKey,
		Commitments:        treeResponse.RootNodeSignatureShares.NodeTxSigningResult.SigningNonceCommitments,
		UserCommitments:    &rootNonceCommitment,
		UserPublicKey:      signingPubkeyBytes,
		UserSignatureShare: userSignatures.Results[nodeJobID].SignatureShare,
	})
	if err != nil {
		return nil, err
	}

	refundSignature, err := frostClient.AggregateFrost(context.Background(), &pbfrost.AggregateFrostRequest{
		Message:            refundTxSighash,
		SignatureShares:    treeResponse.RootNodeSignatureShares.RefundTxSigningResult.SignatureShares,
		PublicShares:       treeResponse.RootNodeSignatureShares.RefundTxSigningResult.PublicKeys,
		VerifyingKey:       verifyingKey,
		Commitments:        treeResponse.RootNodeSignatureShares.RefundTxSigningResult.SigningNonceCommitments,
		UserCommitments:    &refundNonceCommitment,
		UserPublicKey:      signingPubkeyBytes,
		UserSignatureShare: userSignatures.Results[refundJobID].SignatureShare,
	})
	if err != nil {
		return nil, err
	}

	return sparkClient.FinalizeNodeSignatures(context.Background(), &pb.FinalizeNodeSignaturesRequest{
		Intent: pbcommon.SignatureIntent_CREATION,
		NodeSignatures: []*pb.NodeSignatures{
			{
				NodeId:            treeResponse.RootNodeSignatureShares.NodeId,
				NodeTxSignature:   rootSignature.Signature,
				RefundTxSignature: refundSignature.Signature,
			},
		},
	})
}

File: sdks/spark-js-sdk/src/services/deposit.ts
import { schnorr, secp256k1 } from "@noble/curves/secp256k1";
import * as btc from "@scure/btc-signer";
import { p2tr, Transaction } from "@scure/btc-signer";
import { equalBytes, sha256 } from "@scure/btc-signer/utils";
import { SignatureIntent } from "../proto/common";
import {
  Address,
  FinalizeNodeSignaturesResponse,
  GenerateDepositAddressResponse,
  StartTreeCreationResponse,
} from "../proto/spark";
import {
  getP2TRAddressFromPublicKey,
  getSigHashFromTx,
  getTxId,
} from "../utils/bitcoin";
import { subtractPublicKeys } from "../utils/keys";
import { getNetwork } from "../utils/network";
import { proofOfPossessionMessageHashForDepositAddress } from "../utils/proof";
import { createWasmSigningCommitment } from "../utils/signing";
import { WalletConfigService } from "./config";
import { ConnectionManager } from "./connection";
type ValidateDepositAddressParams = {
  address: Address;
  userPubkey: Uint8Array;
};

export type GenerateDepositAddressParams = {
  signingPubkey: Uint8Array;
};

export type CreateTreeRootParams = {
  signingPubKey: Uint8Array;
  verifyingKey: Uint8Array;
  depositTx: Transaction;
  vout: number;
};

const INITIAL_TIME_LOCK = 2000;

export class DepositService {
  private readonly config: WalletConfigService;
  private readonly connectionManager: ConnectionManager;

  constructor(
    config: WalletConfigService,
    connectionManager: ConnectionManager
  ) {
    this.config = config;
    this.connectionManager = connectionManager;
  }

  private async validateDepositAddress({
    address,
    userPubkey,
  }: ValidateDepositAddressParams) {
    if (
      !address.depositAddressProof ||
      !address.depositAddressProof.proofOfPossessionSignature ||
      !address.depositAddressProof.addressSignatures
    ) {
      throw new Error(
        "proof of possession signature or address signatures is null"
      );
    }

    const operatorPubkey = subtractPublicKeys(address.verifyingKey, userPubkey);
    const msg = proofOfPossessionMessageHashForDepositAddress(
      await this.config.signer.getIdentityPublicKey(),
      operatorPubkey,
      address.address
    );

    const taprootKey = p2tr(
      operatorPubkey.slice(1, 33),
      undefined,
      getNetwork(this.config.getNetwork())
    ).tweakedPubkey;

    const isVerified = schnorr.verify(
      address.depositAddressProof.proofOfPossessionSignature,
      msg,
      taprootKey
    );

    if (!isVerified) {
      throw new Error("proof of possession signature verification failed");
    }

    const addrHash = sha256(address.address);
    for (const operator of Object.values(
      this.config.getConfig().signingOperators
    )) {
      if (
        operator.identifier === this.config.getConfig().coodinatorIdentifier
      ) {
        continue;
      }

      const operatorPubkey = operator.identityPublicKey;
      const operatorSig =
        address.depositAddressProof.addressSignatures[operator.identifier];

      const sig = secp256k1.Signature.fromDER(operatorSig);

      const isVerified = secp256k1.verify(sig, addrHash, operatorPubkey);
      if (!isVerified) {
        throw new Error("signature verification failed");
      }
    }
  }

  async generateDepositAddress({
    signingPubkey,
  }: GenerateDepositAddressParams): Promise<GenerateDepositAddressResponse> {
    const sparkClient = await this.connectionManager.createSparkClient(
      this.config.getCoordinatorAddress()
    );

    let depositResp: GenerateDepositAddressResponse;
    try {
      depositResp = await sparkClient.generate_deposit_address({
        signingPublicKey: signingPubkey,
        identityPublicKey: await this.config.signer.getIdentityPublicKey(),
        network: this.config.getNetwork(),
      });
    } catch (error) {
      throw new Error(`Error generating deposit address: ${error}`);
    } finally {
      sparkClient.close?.();
    }
    if (!depositResp.depositAddress) {
      throw new Error("No deposit address response from coordinator");
    }

    this.validateDepositAddress({
      address: depositResp.depositAddress,
      userPubkey: signingPubkey,
    });

    return depositResp;
  }

  async createTreeRoot({
    signingPubKey,
    verifyingKey,
    depositTx,
    vout,
  }: CreateTreeRootParams) {
    // Create a root tx
    const rootTx = new Transaction();
    const output = depositTx.getOutput(vout);
    if (!output) {
      throw new Error("No output found in deposit tx");
    }
    const script = output.script;
    const amount = output.amount;
    if (!script || !amount) {
      throw new Error("No script or amount found in deposit tx");
    }

    rootTx.addInput({
      txid: getTxId(depositTx),
      index: vout,
    });

    rootTx.addOutput({
      script,
      amount,
    });

    const rootNonceCommitment =
      await this.config.signer.getRandomSigningCommitment();
    const rootTxSighash = getSigHashFromTx(rootTx, 0, output);

    // Create a refund tx
    const refundTx = new Transaction();
    const sequence = (1 << 30) | INITIAL_TIME_LOCK;
    refundTx.addInput({
      txid: getTxId(rootTx),
      index: 0,
      sequence,
    });

    const refundP2trAddress = getP2TRAddressFromPublicKey(
      signingPubKey,
      this.config.getNetwork()
    );
    const refundAddress = btc
      .Address(getNetwork(this.config.getNetwork()))
      .decode(refundP2trAddress);
    const refundPkScript = btc.OutScript.encode(refundAddress);

    refundTx.addOutput({
      script: refundPkScript,
      amount: amount,
    });

    const refundNonceCommitment =
      await this.config.signer.getRandomSigningCommitment();
    const refundTxSighash = getSigHashFromTx(refundTx, 0, output);

    const sparkClient = await this.connectionManager.createSparkClient(
      this.config.getCoordinatorAddress()
    );

    let treeResp: StartTreeCreationResponse;

    try {
      treeResp = await sparkClient.start_tree_creation({
        identityPublicKey: await this.config.signer.getIdentityPublicKey(),
        onChainUtxo: {
          vout: vout,
          rawTx: depositTx.toBytes(),
          network: this.config.getNetwork(),
        },
        rootTxSigningJob: {
          rawTx: rootTx.toBytes(),
          signingPublicKey: signingPubKey,
          signingNonceCommitment: rootNonceCommitment,
        },
        refundTxSigningJob: {
          rawTx: refundTx.toBytes(),
          signingPublicKey: signingPubKey,
          signingNonceCommitment: refundNonceCommitment,
        },
      });
    } catch (error) {
      sparkClient.close?.();
      throw new Error(`Error starting tree creation: ${error}`);
    }

    if (!treeResp.rootNodeSignatureShares?.verifyingKey) {
      throw new Error("No verifying key found in tree response");
    }

    if (
      !treeResp.rootNodeSignatureShares.nodeTxSigningResult
        ?.signingNonceCommitments
    ) {
      throw new Error("No signing nonce commitments found in tree response");
    }

    if (
      !treeResp.rootNodeSignatureShares.refundTxSigningResult
        ?.signingNonceCommitments
    ) {
      throw new Error("No signing nonce commitments found in tree response");
    }

    if (
      !equalBytes(treeResp.rootNodeSignatureShares.verifyingKey, verifyingKey)
    ) {
      throw new Error("Verifying key does not match");
    }

    const rootSignature = await this.config.signer.signFrost({
      message: rootTxSighash,
      publicKey: signingPubKey,
      privateAsPubKey: signingPubKey,
      verifyingKey,
      selfCommitment: rootNonceCommitment,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult
          .signingNonceCommitments,
      adaptorPubKey: new Uint8Array(),
    });

    const refundSignature = await this.config.signer.signFrost({
      message: refundTxSighash,
      publicKey: signingPubKey,
      privateAsPubKey: signingPubKey,
      verifyingKey,
      selfCommitment: refundNonceCommitment,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.refundTxSigningResult
          .signingNonceCommitments,
      adaptorPubKey: new Uint8Array(),
    });

    const rootAggregate = await this.config.signer.aggregateFrost({
      message: rootTxSighash,
      statechainSignatures:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult.signatureShares,
      statechainPublicKeys:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult.publicKeys,
      verifyingKey: treeResp.rootNodeSignatureShares.verifyingKey,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult
          .signingNonceCommitments,
      selfCommitment: createWasmSigningCommitment(rootNonceCommitment),
      publicKey: signingPubKey,
      selfSignature: rootSignature!,
      adaptorPubKey: new Uint8Array(),
    });

    const refundAggregate = await this.config.signer.aggregateFrost({
      message: refundTxSighash,
      statechainSignatures:
        treeResp.rootNodeSignatureShares.refundTxSigningResult.signatureShares,
      statechainPublicKeys:
        treeResp.rootNodeSignatureShares.refundTxSigningResult.publicKeys,
      verifyingKey: treeResp.rootNodeSignatureShares.verifyingKey,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.refundTxSigningResult
          .signingNonceCommitments,
      selfCommitment: createWasmSigningCommitment(refundNonceCommitment),
      publicKey: signingPubKey,
      selfSignature: refundSignature,
      adaptorPubKey: new Uint8Array(),
    });

    let finalizeResp: FinalizeNodeSignaturesResponse;
    try {
      finalizeResp = await sparkClient.finalize_node_signatures({
        intent: SignatureIntent.CREATION,
        nodeSignatures: [
          {
            nodeId: treeResp.rootNodeSignatureShares.nodeId,
            nodeTxSignature: rootAggregate,
            refundTxSignature: refundAggregate,
          },
        ],
      });
    } catch (error) {
      sparkClient.close?.();
      throw new Error(`Error finalizing node signatures in deposit: ${error}`);
    } finally {
      sparkClient.close?.();
    }

    return finalizeResp;
  }
}
